File: packages/schemas/src/constitution/load.ts
/**
 * Read-only loaders for the shared product constitution JSON and fixtures.
 * Policy values live only under packages/schemas/constitution/; this module
 * never exposes mutation or write APIs for public endpoints.
 *
 * Serverless note: Next/Vercel bundles this module into chunks where
 * `import.meta.url` no longer points at packages/schemas. Prefer resolving via
 * `@repo/schemas/package.json`, then cwd/monorepo fallbacks, then the source
 * layout relative to this file (App Hosting / local).
 */
import { existsSync, readFileSync } from 'node:fs';
import { createRequire } from 'node:module';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  constitutionFixtureSchema,
  productConstitutionSchema,
  type ConstitutionFixture,
  type ProductConstitution,
} from './schema.js';

const require = createRequire(import.meta.url);

/** Prefer DISCOVERY_REPO_ROOT (Cloud Functions upload root) when the bundle inlines this module. */
function resolveSchemasPackageRoot(): string {
  const deployRoot = process.env.DISCOVERY_REPO_ROOT?.trim();
  if (deployRoot !== undefined && deployRoot.length > 0) {
    return join(deployRoot, 'packages', 'schemas');
  }

  try {
    return dirname(require.resolve('@repo/schemas/package.json'));
  } catch {
    // Bundlers may erase the package entry; fall through.
  }

  const cwdCandidates = [
    join(process.cwd(), 'packages', 'schemas'),
    join(process.cwd(), '..', '..', 'packages', 'schemas'),
    join(process.cwd(), '..', 'packages', 'schemas'),
  ];
  for (const candidate of cwdCandidates) {
    if (existsSync(join(candidate, 'constitution', 'policy.v1.json'))) {
      return candidate;
    }
  }

  return join(dirname(fileURLToPath(import.meta.url)), '../..');
}

function constitutionDir(): string {
  return join(resolveSchemasPackageRoot(), 'constitution');
}

export const CONSTITUTION_DIR = constitutionDir();
export const POLICY_V1_PATH = join(constitutionDir(), 'policy.v1.json');
export const CONSTITUTION_SCHEMA_PATH = join(constitutionDir(), 'product-constitution.schema.json');
export const FIXTURES_DIR = join(constitutionDir(), 'fixtures');

const FIXTURE_FILES = {
  included: 'included.json',
  excluded: 'excluded.json',
  disputed: 'disputed.json',
  sparse: 'sparse.json',
  sensitive: 'sensitive.json',
  living_person: 'living-person.json',
} as const;

export type FixtureKind = keyof typeof FIXTURE_FILES;

let cachedPolicy: ProductConstitution | undefined;

function readJsonFile(path: string): unknown {
  return JSON.parse(readFileSync(path, 'utf8')) as unknown;
}

/** Load and validate the active product constitution (cached, read-only).  */
export function loadProductConstitution(): ProductConstitution {
  if (cachedPolicy) {
    return cachedPolicy;
  }
  const policyPath = join(constitutionDir(), 'policy.v1.json');
  const parsed = productConstitutionSchema.parse(readJsonFile(policyPath));
  cachedPolicy = Object.freeze(structuredClone(parsed)) as ProductConstitution;
  return cachedPolicy;
}

/** Return the policy version string from the loaded constitution.  */
export function getPolicyVersion(policy: ProductConstitution = loadProductConstitution()): string {
  return policy.policyVersion;
}

/** Load a named constitution fixture used by evaluation tests.  */
export function loadConstitutionFixture(kind: FixtureKind): ConstitutionFixture {
  const fileName = FIXTURE_FILES[kind];
  return constitutionFixtureSchema.parse(
    readJsonFile(join(constitutionDir(), 'fixtures', fileName)),
  );
}

/** Load all constitution fixtures (included, excluded, disputed, sparse, sensitive, living-person).  */
export function loadAllConstitutionFixtures(): Record<FixtureKind, ConstitutionFixture> {
  return {
    included: loadConstitutionFixture('included'),
    excluded: loadConstitutionFixture('excluded'),
    disputed: loadConstitutionFixture('disputed'),
    sparse: loadConstitutionFixture('sparse'),
    sensitive: loadConstitutionFixture('sensitive'),
    living_person: loadConstitutionFixture('living_person'),
  };
}

/** Test helper: clear the in-memory policy cache.  */
export function resetProductConstitutionCache(): void {
  cachedPolicy = undefined;
}

File: packages/constitution/src/black_book_constitution/load.py
"""
Load and validate the shared product constitution JSON artifacts.

Values and structure live under packages/schemas/constitution/. This module is
read-only: it caches a validated policy document and never writes policy state.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator

# packages/constitution/src/black_book_constitution -> packages/
_PACKAGES_ROOT = Path(__file__).resolve().parents[3]
CONSTITUTION_DIR = _PACKAGES_ROOT / "schemas" / "constitution"
POLICY_V1_PATH = CONSTITUTION_DIR / "policy.v1.json"
CONSTITUTION_SCHEMA_PATH = CONSTITUTION_DIR / "product-constitution.schema.json"
FIXTURES_DIR = CONSTITUTION_DIR / "fixtures"

_FIXTURE_FILES = {
    "included": "included.json",
    "excluded": "excluded.json",
    "disputed": "disputed.json",
    "sparse": "sparse.json",
    "sensitive": "sensitive.json",
    "living_person": "living-person.json",
}

_cached_policy: dict[str, Any] | None = None


def _read_json(path: Path) -> Any:
    with path.open(encoding="utf-8") as handle:
        return json.load(handle)


def load_product_constitution() -> dict[str, Any]:
    """Load and validate policy.v1.json against the shared JSON Schema."""
    global _cached_policy
    if _cached_policy is not None:
        return _cached_policy

    schema = _read_json(CONSTITUTION_SCHEMA_PATH)
    policy = _read_json(POLICY_V1_PATH)
    Draft202012Validator(schema).validate(policy)
    _cached_policy = policy
    return _cached_policy


def get_policy_version(policy: dict[str, Any] | None = None) -> str:
    """Return the active constitution policyVersion."""
    document = policy if policy is not None else load_product_constitution()
    return str(document["policyVersion"])


def load_constitution_fixture(kind: str) -> dict[str, Any]:
    """Load a named constitution fixture used by evaluation tests."""
    try:
        file_name = _FIXTURE_FILES[kind]
    except KeyError as exc:
        raise KeyError(f"Unknown fixture kind: {kind}") from exc
    return _read_json(FIXTURES_DIR / file_name)


def load_all_constitution_fixtures() -> dict[str, dict[str, Any]]:
    """Load all required constitution fixtures."""
    return {kind: load_constitution_fixture(kind) for kind in _FIXTURE_FILES}


def reset_product_constitution_cache() -> None:
    """Clear the in-memory policy cache (tests only)."""
    global _cached_policy
    _cached_policy = None
